src/gf_nf4.h

```diff
@@ -507,7 +507,19 @@ void NF4<__uint128_t>::hadamard_mul(int n, __uint128_t* x, __uint128_t* y);
 template <typename T>
 void NF4<T>::hadamard_mul(int n, T* x, T* y)
 {
-    return;
+    int i;
```

Comment:
You could reduce the scope of `i` by declaring it in the loop directly (same remark for `add`)

Reply:
in fact `i` appears in two loops so I keep it here.

Reply:
But it's not the same `i` (it is set to 0 for each loop), it makes more sense to declare it inline

for (int i = 0; …)

It help the reader (and maybe the compiler) to see the scope of the variable.

Reply:
it's done